File: code/api_python/benchmarkparser/benchmark_parser.py
class BenchmarkParser: 
    def __init__(self):
        pass

    def parse_benchmark(self, path: str):
        file_content = []

        try:
            file = open(path, 'r')
            file_content = file.readlines()
        except Exception as exception: print(exception) 

        info = file_content[0].split(' ')
        n_machines = int(info[1])
        n_overall_operations = 0
        lines = list(range(file_content.count() - 1))
        for i in range(1, file_content.count()):
            line = file_content[i].split(' ')
            lines[i - 1] = line
            n_overall_operations += int(line[0])

        durations = [n_overall_operations, n_machines]
        operation_index = 0
        job_sequence = range(n_overall_operations)

        for i in range(1, len(lines)):
            line = lines[i-1]
            n_operations = int(line[0])
            index = 1
            for j in range(0, n_operations):
                job_sequence[operation_index] = i-1
                n_options = int(line[index])
                index +=1
                for k in range(0, n_options):
                    machine = int(line[index])
                    index += 1
                    duration = int(line[index])
                    index += 1
                    durations[operation_index, machine - 1] = duration
                operation_index += 1
        return Encoding(durations, job_sequence)

        lines = [line.split(' ') for line in file_content[1:]]
        for line in lines:
            n_overall_operations += int(line[0])

        durations = [[0 for _ in range(n_machines)] for _ in range(n_overall_operations)]
        job_sequence = [0] * n_overall_operations
        operation_index = 0
        for i, line in enumerate(lines):
            n_operations = int(line[0])
            index = 1
            for _ in range(n_operations):
                job_sequence[operation_index] = i
                n_options = int(line[index])
                index += 1
                for _ in range(n_options):
                    machine = int(line[index]) - 1
                    duration = int(line[index+1])
                    durations[operation_index][machine] = duration
                    index += 2
                operation_index += 1

        return Encoding(durations, job_sequence)

File: code/upgrades/CS_Translation/BenchmarkParsing/BenchmarkParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BenchmarkParsing
{
    public class BenchmarkParser
    {
        public BenchmarkParser()
        {

        }

        public Encoding ParseBenchmark(string path)
        {
            List<string> fileContent = new List<string>();
            try
            {
                StreamReader sr = new StreamReader(path);
                string line = sr.ReadLine();
                while (line != null)
                {
                    fileContent.Add(line);
                    line = sr.ReadLine();
                }
                sr.Close();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
            string[] info = fileContent[0].Split(' ');

            int nMachines;
            Int32.TryParse(info[1], out nMachines);
            int nOverallOperations = 0;
            string[][] lines = new string[fileContent.Count - 1][];
            for (int i = 1; i < fileContent.Count; ++i)
            {
                string[] line = fileContent[i].Split(' ');
                lines[i - 1] = line;
                nOverallOperations += Int32.Parse(line[0]);
            }
            int[,] durations = new int[nOverallOperations, nMachines];
            int operationIndex = 0;
            int[] jobSequence = new int[nOverallOperations];
            for (int i = 1; i < lines.Length+1; ++i)
            {
                //string[] line = fileContent[i].Split(' ');
                string[] line = lines[i-1];
                int nOperations;
                Int32.TryParse(line[0], out nOperations);
                int index = 1;
                for (int j = 0; j < nOperations; ++j)
                {
                    jobSequence[operationIndex] = i - 1; // NOTE: switching benchmark input to 0 indexing
                    int nOptions;
                    Int32.TryParse(line[index++], out nOptions);
                    for (int k = 0; k < nOptions; ++k)
                    {
                        int machine;
                        Int32.TryParse(line[index++], out machine);
                        int duration;
                        Int32.TryParse(line[index++], out duration);
                        durations[operationIndex, machine - 1] = duration; // NOTE: switching benchmark input to 0 indexing
                    }
                    ++operationIndex;
                }
            }
            return new Encoding(durations, jobSequence);
        }

        public class WorkerBenchmarkParser
        {
            public WorkerEncoding ParseBenchmark(string path)
            {
                List<string> fileContent = new List<string>();
                try
                {
                    StreamReader sr = new StreamReader(path);
                    string line = sr.ReadLine();
                    while (line != null)
                    {
                        fileContent.Add(line);
                        line = sr.ReadLine();
                    }
                    sr.Close();
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                }
                string[] info = fileContent[0].Split(' ');

                int nMachines;
                Int32.TryParse(info[1], out nMachines);
                int nWorkers;
                Int32.TryParse(info[2], out nWorkers);
                int nOverallOperations = 0;
                string[][] lines = new string[fileContent.Count - 1][];
                for (int i = 1; i < fileContent.Count; ++i)
                {
                    string[] line = fileContent[i].Split(' ');
                    lines[i - 1] = line;
                    nOverallOperations += Int32.Parse(line[0]);
                }
                int[,,] durations = new int[nOverallOperations, nMachines, nWorkers];
                int operationIndex = 0;
                int[] jobSequence = new int[nOverallOperations];
                for (int i = 1; i < lines.Length +1; ++i)
                {
                    string[] line = lines[i - 1];
                    int nOperations;
                    Int32.TryParse(line[0], out nOperations);
                    int index = 1;
                    for (int j = 0; j < nOperations; ++j)
                    {
                        jobSequence[operationIndex] = i - 1; // NOTE: switching benchmark input to 0 indexing
                        int nMachineOptions;
                        Int32.TryParse(line[index++], out nMachineOptions);
                        for (int k = 0; k < nMachineOptions; ++k)
                        {
                            int machine;
                            Int32.TryParse(line[index++], out machine);
                            int nWorkerOptions;
                            Int32.TryParse(line[index++], out nWorkerOptions);
                            for(int l = 0; l < nWorkerOptions; ++l)
                            {
                                int worker;
                                Int32.TryParse(line[index++], out worker);
                                int duration;
                                Int32.TryParse(line[index++], out duration);
                                durations[operationIndex, machine - 1, worker-1] = duration; // NOTE: switching benchmark input to 0 indexing
                            }
                        }
                        ++operationIndex;
                    }
                }
                return new WorkerEncoding(durations, jobSequence);
            }
        }
    }
}
